building/trusted_applications.rst

```diff
@@ -299,3 +305,90 @@ sequence:
             return TEE_ERROR_NOT_SUPPORTED;
         }
     }
+
+.. _TASign:
+
+Signing of TAs
+**************
+
+All Trusted Applications are signed and verified by :ref:`optee_os` upon loading. Within the :ref:`optee_os` source is a directory ``keys``. The public parts of the keys in this directory will be compiled into the :ref:`optee_os` binary and the signature of each TA will be verified against this key(s) upon loading.
+
+.. warning::
+
+   :ref:`optee_os` comes with a default **private** key in its source to facilitate easy development, testing, debugging and QA.
+   Never deploy an :ref:`optee_os` binary with this key in production. Instead replace this keys as soon as possible with a public key and keep the private part of the key offline, preferably on an HSM.
+
+.. note::
+
+   Currently only a single key for signing TAs is supported by :ref:`optee_os`.
+
+TAs are signed using the ``sign.py`` script referenced from ``ta/mk/ta_dev_kit.mk`` in :ref:`optee_os`. Its default behavior is to sign a compiled TA binary and attach the signature to form a complete TA for deployment. For **offline** signing, a three-step process is required: In a first step a digest of the compiled binary has to be generated, in the second step this digest is signed offline using the private key and finally in the third step the binary and its signature are stitched together into the full TA.
+
+Offline Signing of TAs
+++++++++++++++++++++++
+
+The TA dev kit does sign an application as last step of the linking process. For example, the file ``ta/arch/arm/link.mk`` in the :ref:`optee_os` source tree contains the statement
+
+.. code-block:: make
+
+    $(q)$(SIGN) --key $(TA_SIGN_KEY) --uuid $(user-ta-uuid) --version 0 \
+	    --in $$< --out $$@
+
+To avoid build errors when signing offline, this make script needs to be adopted. The detailed usage of ``sign.py`` is given below.
+
+Overall, offline signing is done with the following sequence of steps:
+
+0. (Preparation) Generate a 2048 bit RSA key for signing in a secure, offline environment. Extract the public key and copy it to the ``keys`` directory in the :ref:`optee_os` source tree. Adjust ``TA_SIGN_KEY`` for different file/path names. (Copy and) modify the ``link.mk`` file for the default linking step to produce a digest of the TA binary instead of the full TA.
```

Comment:
IMO if the user has to modify `link.mk`, we're doing it wrong. We'd rather want `link.mk` to support both regular ("online"?) and offline signing. In addition to making the life of users easier, it would enable automatic testing by the test suite (optee_test could select offline testing for one of its TAs for instance).

Reply:
Yes. But documentation should also reflect the currents state of the software. So let's keep this in line with https://github.com/OP-TEE/optee_os/pull/3174 and update the doc when the `link.mk` has been updated in https://github.com/OP-TEE/optee_os.

Reply:
> update the doc when the `link.mk` has been updated in https://github.com/OP-TEE/optee_os

Fine with me.

Reply:
@jforissier I added https://github.com/OP-TEE/optee_os/issues/3275 to keep track of you comment and clear the way for this PR to be merged.